src/hotspot/share/gc/serial/cardTableRS.cpp

```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
   }
 }
 
-class VerifyCleanCardClosure: public BasicOopIterateClosure {
-private:
-  HeapWord* _boundary;
-  HeapWord* _begin;
-  HeapWord* _end;
-protected:
+class CheckForUnmarkedOops : public BasicOopIterateClosure {
+  DefNewGeneration* _young_gen;
+  CardTableRS* _card_table;
+  HeapWord*    _unmarked_addr;
+
   template <class T> void do_oop_work(T* p) {
-    HeapWord* jp = (HeapWord*)p;
-    assert(jp >= _begin && jp < _end,
-           "Error: jp " PTR_FORMAT " should be within "
-           "[_begin, _end) = [" PTR_FORMAT "," PTR_FORMAT ")",
-           p2i(jp), p2i(_begin), p2i(_end));
     oop obj = RawAccess<>::oop_load(p);
-    guarantee(obj == nullptr || cast_from_oop<HeapWord*>(obj) >= _boundary,
-              "pointer " PTR_FORMAT " at " PTR_FORMAT " on "
-              "clean card crosses boundary" PTR_FORMAT,
-              p2i(obj), p2i(jp), p2i(_boundary));
+    if (_young_gen->is_in_reserved(obj) &&
```

Comment:
I would prefer, if this code also failed on values outside of the heap apart from `nullptr` for `obj`, but I understand it's not he point of this particular verification code.